### DISCUSSION TITLE
Default collection

### DISCUSSION THREAD

**Discussion by mrusme** [keepassxreboot/keepassxc#8475] [Q&A]
## Summary

I'm currently playing around with https://github.com/99designs/keyring to see if it works with KeePassXC's secret service. It sort of seems to do, in a way. The first thing that I noticed was that `keyring` requires specifying a `ServiceName`, which it then uses to query dbus:

```
method call time=1663272780.210463 sender=:1.66 -> destination=org.freedesktop.secrets serial=6 path=/org/freedesktop/secrets/collection/servicenamehere; interface=org.freedesktop.Secret.Collection; member=SearchItems
```

In KeePassXC the `collection` seems to be the actual database name. I don't know too much about dbus protocols, but to me it seems that the client querying a secret should not have to know how the KeePassXC database is named in order to query it.

I don't know if this makes sense, but it seems that if KeePassXC maps the `/secrets/collections/<NAME>` to the database name, then KeePassXC should also allow setting a database as *default collection*, so that no matter what a dbus client might query, it would end up requesting the *default database*. Just an idea, again, I'm not too much into the dbus internals, so I don't know whether this makes sense or whether it would break other things.

**Comment by droidmonkey** [comment]
Well you are using a raw api with keyring and not a client application. Part of the process of using an application that speaks Secret Service is to enroll your credentials with that service and store that enrollment information for later retrieval. The name of the database is not relevant in this scenario. When you do raw calls then it becomes relevant.

**Comment by michaelk83** [comment]
The Secret Service API spec does not have a concept of "service name". It does have a concept of collections, which is a way to group secrets together. It seems that `keyring` interprets collections as individual services, which is less common. KeePassXC maps collections to databases.

Regardless, when a client application needs to store secrets in Secret Service, it has two options:
1. Work with a collection of its choice, in which case it needs to know and specify the collection name (or "service name" per `keyring` terminology).
2. Work with a well known *alias*, such as the "default" alias. To access that, you would replace `path=/org/freedesktop/secrets/collection/servicenamehere` with `path=/org/freedesktop/secrets/aliases/default`.

When a client application needs to look up secrets (`SearchItems()` call), it can either do that on a specific collection or alias as above, or it can use the global `SearchItems()` on the service level (`path=/org/freedesktop/secrets` and `interface=org.freedesktop.Secret.Service`). The service will find the correct collection where the items are stored, using the provided lookup attributes.

With KeePassXC, there's an additional complication: if the relevant database is locked, it's not able to perform the search: https://github.com/keepassxreboot/keepassxc/issues/4443#issuecomment-599653832 . Though in recent versions it should ask you to unlock the database in that case.

Currently, KeePassXC sets the "default" alias to the [currently active database tab](https://github.com/keepassxreboot/keepassxc/blob/develop/src/fdosecrets/objects/Service.cpp#L82), as long as it has a group exposed to Secret Service.

You can check which paths are available by running the command `qdbus org.freedesktop.secrets`, from a terminal.
A simple way to check which collection an alias is mapped to is by running `qdbus org.freedesktop.secrets /org/freedesktop/secrets/aliases/<name-of-alias> org.freedesktop.Secret.Collection.Label`.

Some relevant links if you want to read deeper:
https://specifications.freedesktop.org/secret-service/latest/ch04.html (on aliases)
https://specifications.freedesktop.org/secret-service/latest/ch03.html (on collection and items)
https://specifications.freedesktop.org/secret-service/latest/ch12.html (on available object paths in Secret Service)
https://specifications.freedesktop.org/secret-service/latest/re01.html (`org.freedesktop.Secret.Service` interface)
https://specifications.freedesktop.org/secret-service/latest/re02.html (`org.freedesktop.Secret.Collection` interface)

**Comment by michaelk83** [comment]
> Currently, KeePassXC sets the "default" alias to the [currently active database tab](https://github.com/keepassxreboot/keepassxc/blob/develop/src/fdosecrets/objects/Service.cpp#L82), as long as it has a group exposed to Secret Service.

@droidmonkey , maybe that should be a setting somewhere? As it is, if there are two databases with exposed groups, switching between them to do something completely unrelated to Sercret Service can cause client apps to access the wrong database.

Furthermore, [the spec suggests](https://specifications.freedesktop.org/secret-service/latest/ch04.html) there should be an interface to manage alias mappings in general:
> For applications like password managers it can be useful to allow the user to configure which collection is associated with which well known alias.

**Reply by droidmonkey** [reply]
Bringing in @Aetf

**Reply by Aetf** [reply]
Sounds good to me. The mapping should probably go to application settings, and either file path or UUID can be used to identify the database.

**Reply by michaelk83** [reply]
#8479

**Comment by mrusme** [comment]
Excuse the delay and thank you very much for the kind and insightful explanations @droidmonkey @michaelk83! 

I will take a closer look at the `keyring` library that I'm using to see if I can make it use an `alias` then and if so, use the `default` alias.

Thank you for your help!
